src/testing/systest/supervisor/systest.zig

```diff
@@ -0,0 +1,241 @@
+    const time_start = std.time.nanoTimestamp();
+
+    var replicas: [replica_count]*LoggedProcess = undefined;
+    for (0..replica_count) |i| {
+        const name = try shell.fmt("replica{d}", .{i});
+        const datafile = try shell.fmt("{s}/1_{d}.tigerbeetle", .{ tmp_dir, i });
+
+        // Format datafile
+        try shell.exec(
+            \\{tigerbeetle} format 
+            \\  --cluster=1
+            \\  --replica={index}
+            \\  --replica-count={replica_count} 
+            \\  {datafile}
+        , .{
+            .tigerbeetle = args.tigerbeetle_executable,
+            .index = i,
+            .replica_count = replica_count,
+            .datafile = datafile,
+        });
+
+        // Start replica
+        const addresses = try shell.fmt("--addresses={s}", .{
+            try comma_separate_ports(shell.arena.allocator(), &replica_ports),
+        });
+        const argv = try shell.arena.allocator().dupe([]const u8, &.{
+            args.tigerbeetle_executable,
+            "start",
+            addresses,
+            datafile,
+        });
+
+        var process = try LoggedProcess.create(allocator, name, argv, .{});
+        errdefer process.destroy();
+
+        replicas[i] = process;
+        try process.start();
+    }
+
+    // Start workload
+    const workload = try start_workload(shell, allocator);
+    errdefer workload.destroy();
+
+    // Set up nemesis (fault injector)
+    var prng = std.rand.DefaultPrng.init(0);
+    const nemesis = try Nemesis.init(allocator, prng.random(), &replicas);
+    defer nemesis.deinit();
+
+    // Let the workload finish by itself, or kill it after we've run for the required duration.
+    // Note that the nemesis is blocking in this loop.
+    const workload_result = term: {
+        while (std.time.nanoTimestamp() - time_start < test_duration_ns) {
+            // Try to do something funky in the nemesis, and if it fails, wait for a while
+            // before trying again.
+            if (!try nemesis.wreak_havoc()) {
+                std.time.sleep(100 * std.time.ns_per_ms);
+            }
+            if (workload.state() == .completed) {
+                log.info("supervisor: workload completed by itself", .{});
+                break :term try workload.wait();
+            }
+        }
+
+        log.info("supervisor: terminating workload due to max duration", .{});
+        break :term try workload.terminate();
+    };
+
+    workload.destroy();
+
+    for (replicas) |replica| {
+        // The nemesis might have terminated the replica and never restarted it,
+        // so we need to check its state.
+        if (replica.state() == .running) {
+            _ = try replica.terminate();
+        }
+        replica.destroy();
+    }
+
+    switch (workload_result) {
+        .Exited => |code| {
+            if (code == 128 + std.posix.SIG.KILL) {
+                log.info("supervisor: workload terminated (SIGKILL) as requested", .{});
```

Comment:
Hm, this is not right I think? 128+signal is the shell thing, but we don't have a shell here, so we have proper .Exited/.Signal thing here.